[libc] add remaining epoll functions, pipe

michaelrj-google opened the pull request:
The epoll_wait functions need the rest of the epoll functions (create,
ctl) to be available to actually test them, as well as pipe to create a
usable file descriptor. This patch adds epoll_create, epoll_create1,
epoll_ctl, and pipe. These have tests, and the tests for epoll_wait,
epoll_pwait, and epoll_pwait2 (currently disabled) are updated to use
these newly available functions.

michaelrj-google committed:
[libc] add remaining epoll functions, pipe

The epoll_wait functions need the rest of the epoll functions (create,
ctl) to be available to actually test them, as well as pipe to create a
usable file descriptor. This patch adds epoll_create, epoll_create1,
epoll_ctl, and pipe. These have tests, and the tests for epoll_wait,
epoll_pwait, and epoll_pwait2 (currently disabled) are updated to use
these newly available functions.

lntue commented on libc/include/llvm-libc-macros/linux/sys-epoll-macros.h:
```diff
@@ -0,0 +1,38 @@
+#ifndef LLVM_LIBC_MACROS_LINUX_SYS_EPOLL_MACROS_H
+#define LLVM_LIBC_MACROS_LINUX_SYS_EPOLL_MACROS_H
+
+#include "llvm-libc-macros/linux/fcntl-macros.h"
```
Use relative path here.

nickdesaulniers commented on libc/include/llvm-libc-macros/linux/sys-epoll-macros.h:
```diff
@@ -0,0 +1,38 @@
+
+#include "llvm-libc-macros/linux/fcntl-macros.h"
+
+// TODO: Do we need to define these?
```
`<linux/eventpoll.h>`?

nickdesaulniers replied:
That's glibc's definition, which indeed seems subtly different from the kernel's (as is noted in that man page).  I think alluding to this issue in this comment would be better than the todo.

nickdesaulniers replied:
It sounds like `epoll_event` is packed on x86_64. Our definition in libc/include/llvm-libc-types/struct_epoll_event.h is missing that.

nickdesaulniers commented on libc/spec/posix.td:
```diff
@@ -548,6 +548,11 @@ def POSIX : StandardSpec<"POSIX"> {
+        FunctionSpec<
+          "pipe",
+          RetValSpec<IntType>,
+          [ArgSpec<IntPtr>] //TODO: make this int[2]
```
Do we not already have support for array params?

nickdesaulniers commented on libc/src/sys/epoll/linux/epoll_create.cpp:
```diff
@@ -0,0 +1,39 @@
+#ifdef SYS_epoll_create
+  int ret = LIBC_NAMESPACE::syscall_impl<int>(SYS_epoll_create, size);
+#elif defined(SYS_epoll_create1)
+  // TODO: Silence warning for size being unused.
```
add `[[gnu::maybe_used]]` on the parameter. `(void) size;` only works for clang but not yet for GCC.

nickdesaulniers commented on libc/src/unistd/linux/pipe.cpp:
```diff
@@ -0,0 +1,33 @@
+namespace LIBC_NAMESPACE {
+
+LLVM_LIBC_FUNCTION(int, pipe, (int pipefd[2])) {
+#ifdef SYS_pipe
+  int ret = LIBC_NAMESPACE::syscall_impl<int>(SYS_pipe,
+                                              reinterpret_cast<long>(pipefd));
+#elif defined(SYS_pipe2)
```
again, if pipe2 is newer, than shouldn't we be checking for then using that first? man page says linux 2.6.27 (ancient)

nickdesaulniers commented on libc/test/src/sys/epoll/linux/epoll_create1_test.cpp:
```diff
@@ -0,0 +1,32 @@
+using namespace LIBC_NAMESPACE::testing::ErrnoSetterMatcher;
+
+TEST(LlvmLibcEpollCreate1Test, Basic) {
+  int fd = LIBC_NAMESPACE::epoll_create1(0);
+  ASSERT_GT(fd, 0);
```
Maybe https://android.googlesource.com/platform/bionic/+/refs/heads/main/tests/sys_epoll_test.cpp has some inspiration, like testing the CLOEXEC bit.

nickdesaulniers commented on libc/test/src/sys/epoll/linux/epoll_create_test.cpp:
```diff
@@ -0,0 +1,22 @@
+using namespace LIBC_NAMESPACE::testing::ErrnoSetterMatcher;
+
+TEST(LlvmLibcEpollCreateTest, Basic) {
+  int fd = LIBC_NAMESPACE::epoll_create(1);
```
https://android.googlesource.com/platform/bionic/+/refs/heads/main/tests/sys_epoll_test.cpp has a test that input `0` fails. Consider adding a test for that.

nickdesaulniers commented on libc/test/src/sys/epoll/linux/epoll_create1_test.cpp:
```diff
@@ -0,0 +1,32 @@
+#include "test/UnitTest/ErrnoSetterMatcher.h"
+#include "test/UnitTest/Test.h"
+
+#include <sys/epoll.h>
```
Why do you pull in this system header, for this test and the ones below?

nickdesaulniers replied:
So does your newly added libc/include/llvm-libc-macros/linux/sys-epoll-macros.h, why not use that for this test, rather than the system header?

nickdesaulniers reviewed (commented):
In the future, consider breaking these function implementations up into distinct PRs.  That way, we can make incremental progress on landing pieces.  Things flagged in code review don't block landing unrelated functions.

michaelrj-google committed:
address comments

nickdesaulniers commented on libc/src/sys/epoll/epoll_ctl.h:
```diff
@@ -0,0 +1,25 @@
+#ifndef LLVM_LIBC_SRC_SYS_EPOLL_EPOLL_CTL_H
+#define LLVM_LIBC_SRC_SYS_EPOLL_EPOLL_CTL_H
+
+// TODO: Use this include once the include headers are also using quotes.
```
Copying notes from an in person discussion.

I think it will help us avoid including system headers (which we have a linter check for, which is the largest blocker to getting the linter re-enabled in post- then pre- submit) which may make use of hazards like `#include_next` (which was a repeated issue with `stdint.h`) if we provide definitions which are compatible with the systems we support. Those systems have well defined ABIs that are unlikely to change (without massive heads up to native application developers).  And when these definitions differ between platforms, we have control over which mutually exclusive definition we build against.

So, for this PR, I think we can avoid `#include <sys/epoll.h>` by improving our definition of `struct epoll_event` in libc/include/llvm-libc-types/struct_epoll_event.h to denote that it's packed for x86_64, then I think we can use our libc/include/llvm-libc-types/struct_epoll_event.h and avoid <sys/epoll.h> or the curious <linux/eventpoll.h>.

nickdesaulniers commented on libc/src/sys/epoll/linux/epoll_create.cpp:
```diff
@@ -0,0 +1,38 @@
+#include "src/__support/OSUtil/syscall.h" // For internal syscall function.
+#include "src/__support/common.h"
+#include "src/errno/libc_errno.h"
+#include <sys/syscall.h> // For syscall numbers.
```
What is libc/config/linux/syscall_numbers.h.inc (which also looks like it defines SYS_epoll_create and SYS_epoll_create1)?  I wonder if we should be using that, too, for the same reasons for not using `<sys/epoll.h>` listed above?  Looks like that generates a sys/syscall.h, but I'm guessing that't only for fullbuild mode and not overlay mode?

michaelrj-google replied:
It's marked as a generated header right now, but it doesn't appear to actually need to be. We should look into moving it to /include/ at some point.

michaelrj-google committed:
redo includes

nickdesaulniers reviewed (approved):
Don't forget to pack the `struct epoll_event` for x86_64 before you land this.

michaelrj-google committed:
pack the struct, adjust comment in macro file

nickdesaulniers commented on libc/src/sys/epoll/epoll_ctl.h:
```diff
@@ -0,0 +1,25 @@
+// #include "include/llvm-libc-types/sigset_t.h"
+// #include "include/llvm-libc-types/struct_epoll_event.h"
+
+#include <sys/epoll.h>
```
ah, this still needs to be removed

nickdesaulniers commented on libc/include/llvm-libc-types/struct_epoll_event.h:
```diff
@@ -11,7 +11,11 @@
 
 #include <llvm-libc-types/struct_epoll_data.h>
```
I think this needs to be adjusted to "struct_epoll_data.h", based on the build failures.
https://buildkite.com/llvm-project/github-pull-requests/builds/46591#018e348b-db56-4d69-a0a2-ca7d6daf9f76

michaelrj-google committed:
clean up includes some more

nickdesaulniers commented on libc/src/sys/epoll/epoll_pwait.h:
```diff
@@ -9,11 +9,11 @@
-// TODO: Use this include once the include headers are also using quotes.
+// TODO: use our internal sigset_t type (issue #86034)
 // #include "include/llvm-libc-types/sigset_t.h"
-// #include "include/llvm-libc-types/struct_epoll_event.h"
+#include <signal.h>
 
-#include <sys/epoll.h>
+#include "include/llvm-libc-types/struct_epoll_event.h"
```
Should this be using `struct epoll_event *` for `events`, rather than `epoll_event*` for `events` param?

Then we don't even need any forward declarations. TIL that in C++ not having the `struct`/(or probably `class`) keyword(s) doesn't allow you to have an incomplete or opaque type.

Example: https://godbolt.org/z/nfE8K7Whd

There's other cases of this above, like in libc/src/sys/epoll/epoll_ctl.h.

michaelrj-google committed:
Rebase and use the proper header indirection system

michaelrj-google commented on libc/test/src/sys/epoll/linux/epoll_pwait_test.cpp:
```diff
@@ -5,16 +5,45 @@
+
+  ASSERT_THAT(LIBC_NAMESPACE::pipe(pipefd), Succeeds());
+
+  epoll_event event{.events = EPOLLOUT, .data = {.fd = pipefd[0]}};
```
for simplicity I moved these to initialize on multiple lines.

michaelrj-google committed:
avoid designated initializers

michaelrj-google committed:
re-comment out epoll_pwait2

The syscall isn't available on all relevant platforms yet, so it needs
to be disabled for now.
